ompi/runtime/ompi_mpi_init.c

```diff
@@ -384,29 +385,37 @@ int ompi_mpi_init(int argc, char **argv, int requested, int *provided)
 
     ompi_hook_base_mpi_init_top(argc, argv, requested, provided);
 
-    /* Ensure that we were not already initialized or finalized.
-
-       This lock is held for the duration of ompi_mpi_init() and
-       ompi_mpi_finalize().  Hence, if we get it, then no other thread
-       is inside the critical section (and we don't have to check the
-       *_started bool variables). */
-    opal_atomic_rmb();
-    int32_t state = ompi_mpi_state;
+    /* Ensure that we were not already initialized or finalized. */
+    int32_t expected = OMPI_MPI_STATE_NOT_INITIALIZED;
+    int32_t desired  = OMPI_MPI_STATE_INIT_STARTED;
+    opal_atomic_wmb();
+    if (!opal_atomic_compare_exchange_strong_32(&ompi_mpi_state, &expected,
+                                                desired)) {
+        // If we failed to atomically transition ompi_mpi_state from
+        // NOT_INITIALIZED to INIT_STARTED, then someone else already
+        // did that, and we should return.
+        if (expected >= OMPI_MPI_STATE_FINALIZE_STARTED) {
+            opal_show_help("help-mpi-runtime.txt",
+                           "mpi_init: already finalized", true);
```

Comment:
that would be heresy ;)

Reply:
Correct: MPI defines that you are only allowed to call MPI_INIT / MPI_FINALIZE once.

We added this `ompi_mpi_init()` ability to "re-init" just for OSHMEM (I put it in quotes because it's not actually re-initializing -- it's just returning safely if MPI is *already* initialized... but not yet finalized).

But we can't help you if MPI has already been finalized -- there's way too much infrastructure that has been shut down, and `ompi_mpi_init()` was not designed to *actually* re-initialize.  MPI sessions may change this reality, but that's still a ways off.  And definitely outside the scope of this PR.  😄